mivideo/mivideo_extract.py

```diff
@@ -0,0 +1,134 @@
+# -*- coding: utf-8 -*-
+'''
+Module for setting up and running the MiVideo data extract.
+'''
+
+import logging
+import os
+import time
+from typing import Union, Sequence, Dict
+
+import pandas as pd
+from google.cloud import bigquery
+from google.oauth2 import service_account
+from sqlalchemy.engine import ResultProxy
+
+import mivideo.queries as queries
+from db.db_creator import DBCreator
+from environ import ENV, CONFIG_DIR
+from vocab import ValidDataSourceName
+
+logger = logging.getLogger(__name__)
+
+SHAPE_ROWS: int = 0  # Index of row count in DataFrame.shape() array
+
+
+class MiVideoExtract(object):
+    '''
+    Initialize the MiVideo data extract process by instantiating the ``MiVideoExtract`` class,
+    then invoke its ``run()`` method.
+
+    For example, ``MiVideoExtract().run()``.
+    '''
+
+    def __init__(self):
+        self.udpKeyFilePath: str = os.path.join(CONFIG_DIR, ENV.get('mivideo', {}).get(
+            'service_account_json_filename'))
+        logger.debug(f'udpKeyFilePath: "{self.udpKeyFilePath}"')
+
+        self.credentials: service_account.Credentials = \
+            service_account.Credentials.from_service_account_file(
+                self.udpKeyFilePath,
+                scopes=['https://www.googleapis.com/auth/cloud-platform'],
+            )
+
+        self.udpDb: bigquery.Client = bigquery.Client(
+            credentials=self.credentials,
+            project=self.credentials.project_id,
+        )
+
+        logger.info(f'Connected to BigQuery project: "{self.udpDb.project}"')
+
+        DB_PARAMS = ENV['INVENTORY_DB']
+        APPEND_TABLE_NAMES = ENV.get('APPEND_TABLE_NAMES', [
+            'job_run', 'data_source_status', 'mivideo_media_started_hourly',
+            'mivideo_media_creation'])
+
+        self.appDb: DBCreator = DBCreator(DB_PARAMS, APPEND_TABLE_NAMES)
+        self.appDb.set_up()
+
+    def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[str, None]:
+        lastTime: Union[str, None]
+
+        try:
+            sql: str = f'select max(t.{tableColumnName}) from {tableName} t'
+            result: ResultProxy = self.appDb.engine.execute(sql)
+            lastTime = str(result.fetchone()[0])
+        except(Exception):
+            lastTime = None
+
+        return lastTime
+
+    def run(self) -> Sequence[Dict[str, Union[ValidDataSourceName, pd.Timestamp]]]:
+        """
+        :return: Sequence of result dictionaries, with ValidDataSourceName and last run timestamp
+        """
+
+        tableName = 'mivideo_media_started_hourly'
+
+        logger.info(f'"{tableName}" - Starting procedure...')
+
+        lastTime: Union[str, None] = self._readTableLastTime(tableName, 'event_time_utc_latest')
```

Comment:
I'm okay with this for now, but if you abstract this more to the point where it's accepting arbitrary input (from `ENV` perhaps), we should probably verify the `tableName` is valid (perhaps using `self.appDB.get_table_names()`.